File: app/src/main/java/com/aivoice/translate/Activity/VoiceActivity.java
package com.aivoice.translate.Activity;

import android.content.Intent;
import android.os.Bundle;
import android.speech.RecognizerIntent;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;

import com.aivoice.translate.AdsUtils.FirebaseADHandlers.AdUtils;
import com.aivoice.translate.AdsUtils.Interfaces.AppInterfaces;
import com.aivoice.translate.AdsUtils.Utils.Constants;
import com.aivoice.translate.R;
import com.aivoice.translate.base.BaseActivity;
import com.aivoice.translate.translateApi.TranslateApi;
import com.aivoice.translate.utils.Utils;

import java.util.ArrayList;
import java.util.Locale;
import java.util.Objects;

import butterknife.BindView;
import butterknife.OnClick;

import static com.aivoice.translate.utils.Utils.WHATIS;

public class VoiceActivity extends BaseActivity {
    @BindView(R.id.mETInput)
    EditText mETInput;
    @BindView(R.id.mTxtResult)
    TextView mTxtResult;
    @BindView(R.id.mTxtInput)
    TextView mTxtInput;
    @BindView(R.id.mTxtFromLang)
    TextView mTxtFromLang;
    @BindView(R.id.mTxtToLang)
    TextView mTxtToLang;
    @BindView(R.id.mIVClear)
    ImageView mIVClear;
    @BindView(R.id.mIVRecordVoice)
    ImageView mIVRecordVoice;
    @BindView(R.id.mIVText)
    ImageView mIVText;
    private String mLanguageCodeFrom = "en";
    private String mLanguageCodeTo = "en";
    boolean IsFrom = true;
    public static VoiceActivity voiceActivity;

    public static VoiceActivity getInstance() {
        return voiceActivity;
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_voice;
    }

    @Override
    protected void initView(Bundle savedInstanceState) {
        AdUtils.showNativeAd(VoiceActivity.this, Constants.adsJsonPOJO.getParameters().getNative_id().getDefaultValue().getValue(), (LinearLayout) findViewById(R.id.native_ads1), true);
        voiceActivity = this;
        mTxtInput.setText("Click Microphone for recording your voice");
        mTxtInput.setVisibility(View.VISIBLE);
        mETInput.setVisibility(View.GONE);
        mIVRecordVoice.setVisibility(View.VISIBLE);
        mIVText.setVisibility(View.GONE);
        mIVClear.setVisibility(View.GONE);

//        AdUtils.showNativeAd(StartActivity.this, Constants.adsJsonPOJO.getParameters().getNative_id().getDefaultValue().getValue(), findViewById(R.id.native_ads), true);
    }

    @OnClick({R.id.mIVBack, R.id.mIVRecordVoice, R.id.mIVTrans, R.id.mRLTranslate, R.id.mTxtFromLang, R.id.mTxtToLang})
    public void onViewClicked(View view) {
        switch (view.getId()) {
            case R.id.mIVBack:
                AdUtils.showInterstitialAd(VoiceActivity.this, new AppInterfaces.InterStitialADInterface() {
                    @Override
                    public void adLoadState(boolean isLoaded) {
                        VoiceActivity.super.onBackPressed();
                    }
                });
                break;
            case R.id.mIVRecordVoice:
                mRecordVoice();
                break;
            case R.id.mIVTrans:
                mTransLanguage();
                break;
            case R.id.mRLTranslate:
                mTranslateText();
                break;
            case R.id.mTxtFromLang:
                IsFrom = true;
                WHATIS="voice";
                startActivity(new Intent(voiceActivity, LanguageActivity.class));
                break;
            case R.id.mTxtToLang:
                IsFrom = false;
                WHATIS="voice";
                startActivity(new Intent(voiceActivity, LanguageActivity.class));
                break;

        }
    }

    private void mTransLanguage() {
        String mTmp = mLanguageCodeFrom;
        mLanguageCodeFrom = mLanguageCodeTo;
        mLanguageCodeTo = mTmp;

        String mLanTmp= mTxtFromLang.getText().toString();
        mTxtFromLang.setText(mTxtToLang.getText().toString());
        mTxtToLang.setText(mLanTmp);
    }

    private void mTranslateText() {
        Utils.hideKeyboard(mETInput);
        TranslateApi translate = new TranslateApi();
        translate.setOnTranslationCompleteListener(new TranslateApi.OnTranslationCompleteListener() {
            @Override
            public void onStartTranslation() {
                // here you can perform initial work before translated the text like displaying progress bar
            }

            @Override
            public void onCompleted(String text) {
                // "text" variable will give you the translated text
                mTxtResult.setText(text);
            }

            @Override
            public void onError(Exception e) {

            }
        });
        translate.execute(mTxtInput.getText().toString(), mLanguageCodeFrom, mLanguageCodeTo);
    }

    public void mRecordVoice() {
        Intent intent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, Locale.getDefault());
        intent.putExtra(RecognizerIntent.EXTRA_PROMPT, "Speak to text");

        try {
            startActivityForResult(intent, 22);
        } catch (Exception e) {
            Toast.makeText(VoiceActivity.this, " " + e.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 22) {
            if (resultCode == RESULT_OK && data != null) {
                ArrayList<String> result = data.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS);
                mTxtInput.setText(Objects.requireNonNull(result).get(0));
            }
        }
    }

    public void mGetSelctedLang(String lang, String str) {
        if (IsFrom) {
            mLanguageCodeFrom = str;
            mTxtFromLang.setText("" + lang);
        } else {
            mLanguageCodeTo = str;
            mTxtToLang.setText("" + lang);
        }
    }

    @Override
    public void onBackPressed() {
        AdUtils.showInterstitialAd(VoiceActivity.this, new AppInterfaces.InterStitialADInterface() {
            @Override
            public void adLoadState(boolean isLoaded) {
                VoiceActivity.super.onBackPressed();
            }
        });
    }
}

File: app/src/main/java/com/texttranslate/voiceimage/VoiceActivity.kt
package com.texttranslate.voiceimage

import android.content.Intent
import android.os.Bundle
import android.speech.RecognizerIntent
import android.view.View
import android.widget.Toast
import com.texttranslate.voiceimage.base.PermissionHelperActivity
import com.texttranslate.voiceimage.databinding.ActivityVoiceBinding
import com.texttranslate.voiceimage.interfaces.OnItemClickListener
import com.texttranslate.voiceimage.interfaces.OnTranslationCompleteListener
import com.texttranslate.voiceimage.translateApi.TranslateApi
import com.texttranslate.voiceimage.utils.Utils
import java.util.Locale

class VoiceActivity : PermissionHelperActivity(), View.OnClickListener, OnItemClickListener {
    private var mLanguageCodeFrom: String? = "en"
    private var mLanguageCodeTo: String? = "en"
    var IsFrom = true
    var voiceActivity: VoiceActivity? = null
    private var binding: ActivityVoiceBinding? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityVoiceBinding.inflate(
            layoutInflater
        )
        setContentView(binding!!.getRoot())
        voiceActivity = this
        binding!!.mTxtInput.text = "Click Microphone for recording your voice"
        binding!!.mTxtInput.visibility = View.VISIBLE
        binding!!.mETInput.visibility = View.GONE
        binding!!.mIVRecordVoice.setVisibility(View.VISIBLE)
        binding!!.mIVText.setVisibility(View.GONE)
        binding!!.mIVClear.setVisibility(View.GONE)
        binding!!.mIVBack.setOnClickListener(this)
        binding!!.mIVRecordVoice.setOnClickListener(this)
        binding!!.mIVTrans.setOnClickListener(this)
        binding!!.mRLTranslate.setOnClickListener(this)
        binding!!.mTxtFromLang.setOnClickListener(this)
        binding!!.mTxtToLang.setOnClickListener(this)
    }

    var dialog: LanguageDialog? = null
    override fun onClick(v: View) {
        if (v === binding!!.mIVBack) {
            super@VoiceActivity.finish()
        } else if (v === binding!!.mIVRecordVoice) {
            mRecordVoice()
        } else if (v === binding!!.mIVTrans) {
            mTransLanguage()
        } else if (v === binding!!.mRLTranslate) {
            mTranslateText()
        } else if (v === binding!!.mTxtFromLang) {
            IsFrom = true
            Utils.WHATIS = "voice"
            dialog = LanguageDialog(voiceActivity!!, this)
            dialog!!.show()
        } else if (v === binding!!.mTxtToLang) {
            IsFrom = false
            Utils.WHATIS = "voice"
            dialog = LanguageDialog(voiceActivity!!, this)
            dialog!!.show()
        }
    }

    private fun mTransLanguage() {
        val mTmp = mLanguageCodeFrom
        mLanguageCodeFrom = mLanguageCodeTo
        mLanguageCodeTo = mTmp
        val mLanTmp = binding!!.mTxtFromLang.getText().toString()
        binding!!.mTxtFromLang.text = binding!!.mTxtToLang.getText().toString()
        binding!!.mTxtToLang.text = mLanTmp
    }

    private fun mTranslateText() {
        Utils.hideKeyboard(binding!!.mETInput)
        val translate = TranslateApi()
        translate.setOnTranslationCompleteListener(object : OnTranslationCompleteListener {
            override fun onStartTranslation() {
                // here you can perform initial work before translated the text like displaying progress bar
            }

            override fun onCompleted(text: String?) {
                // "text" variable will give you the translated text
                binding!!.mTxtResult.text = text
            }

            override fun onError(e: Exception?) {}
        })
        translate.execute(
            binding!!.mTxtInput.getText().toString(),
            mLanguageCodeFrom,
            mLanguageCodeTo
        )
    }

    fun mRecordVoice() {
        val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH)
        intent.putExtra(
            RecognizerIntent.EXTRA_LANGUAGE_MODEL,
            RecognizerIntent.LANGUAGE_MODEL_FREE_FORM
        )
        intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, Locale.getDefault())
        intent.putExtra(RecognizerIntent.EXTRA_PROMPT, "Speak to text")
        try {
            startActivityForResult(intent, 22)
        } catch (e: Exception) {
            Toast.makeText(this@VoiceActivity, " " + e.message, Toast.LENGTH_SHORT).show()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == 22) {
            if (resultCode == RESULT_OK && data != null) {
                val result = data.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS)
                binding!!.mTxtInput.text = result!![0]
            }
        }
    }

    override fun mGetSelctedLang(lang: String?, str: String?) {
        if (IsFrom) {
            mLanguageCodeFrom = str
            binding!!.mTxtFromLang.text = "" + lang
        } else {
            mLanguageCodeTo = str
            binding!!.mTxtToLang.text = "" + lang
        }
        dialog!!.dismiss()
    }
}